### PR TITLE
Obj index update

### PR DESCRIPTION
**Pull request by movestill**
Step functions and automation for id indexing.

Note, submodules haven't been updated yet and this is an incremental PR. These branches have lived too long and should be merged back to integration.

Accompanying PRs:
* https://github.com/jhuapl-boss/boss-tools/pull/4
* https://github.com/jhuapl-boss/boss/pull/10
* https://github.com/jhuapl-boss/spdb/pull/3

### CODE REVIEW

File: cloud_formation/configs/dynamolambda.py [hunk 1]
```diff
@@ -96,7 +96,8 @@ def create_config(session, domain):
                           "index.handler"),
                       timeout=120,
                       memory=128,
-                      runtime="nodejs6.10")
+                      runtime="nodejs6.10",
+                      reserved_executions=1)
```
Comment (dkleissa) [inline cloud_formation/configs/dynamolambda.py]: is this nodejs lambda the 3rd party dynamodb autoscaler? The new integrated dynamodb autoscaling capability works pretty well and may simplify deployment. It's worth considering switching to it.

File: cloud_formation/configs/idindexing.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+    config.add_lambda(
+        "indexLoadIdsFromS3Lambda",
+        names.index_load_ids_from_s3_lambda,
+        Ref("LambdaCacheExecutionRole"),
+        s3=(aws.get_lambda_s3_bucket(session),
+            "multilambda.{}.zip".format(domain),
+            "load_ids_from_s3_lambda.handler"),
+        timeout=120,
+        memory=128,
```
Comment (dkleissa) [inline cloud_formation/configs/idindexing.py]: 👏 👏 👏 I see individual timeout and memory settings now on lambda creation now

File: cloud_formation/stepfunctions/index_fanout_enqueue_cuboids.hsd [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    result: '$.fanout_params'
+    data:
+        {
+            "max_concurrent": 3,
+            "rampup_delay": 4,
+            "rampup_backoff": 0.5,
+            "status_delay": 1
+        }
+
+while '$.finished' == False:
```
Comment (derek-pryor) [inline cloud_formation/stepfunctions/index_fanout_enqueue_cuboids.hsd]: $.finished is not listed in the input arguments comment

File: salt_stack/salt/lambda-dev/files/makedomainenv [hunk 4]
```diff
@@ -57,6 +57,13 @@ else
    cp makefile_LINUX makefile
    make
 
+   cd ~/virtualenvs/$1/lib/python3.6/site-packages
+   cd heaviside.git
```
Comment (derek-pryor) [inline salt_stack/salt/lambda-dev/files/makedomainenv]: Why not pip install?

File: salt_stack/salt/lambda-dev/files/makedomainenv [hunk 5]
```diff
@@ -68,18 +75,33 @@ else
...
-   chmod -R 077 *
+   chmod -R 0777 *
    zip -r9 ~/lambdazips/multilambda.${1}.zip *
 
    cd ~/virtualenvs/${1}/lib64/atlas
    zip -r9 ~/lambdazips/multilambda.${1}.zip *
 
    cd ~/virtualenvs/${1}/local/lib/python3.6/site-packages/lambda
    zip -g ~/lambdazips/multilambda.${1}.zip lambda_loader.py
+   zip -g ~/lambdazips/multilambda.${1}.zip write_s3_index_lambda.py
```
Comment (derek-pryor) [inline salt_stack/salt/lambda-dev/files/makedomainenv]: would adding `*.py` work better?

File: bin/update_lambda_fcn.py [hunk 6]
```diff
@@ -35,6 +35,7 @@
 from lib import zip
```
Comment (derek-pryor) [inline bin/update_lambda_fcn.py]: Does it make sense to combind all of the lambda scripts into a single one? Just wondering about overpopulating the bin/ directory.

File: cloud_formation/configs/idindexing.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+    """Build multilambda zip file and put in S3."""
+    bucket = aws.get_lambda_s3_bucket(session)
+    load_lambdas_on_s3(session, domain, bucket)
+
+
+def post_init(session, domain):
+    """Create step functions."""
+    names = AWSNames(domain)
+
+    sfn.create(
```
[the remark this answers is not in the document]
Comment (movestill) [reply cloud_formation/configs/idindexing.py]: Save for possible incorporation later? Figure we'd want to update activities at the same time.

File: bin/update_lambda_fcn.py (hunk 6 above)
Comment (movestill) [reply bin/update_lambda_fcn.py]: Potentially. Future task?

File: salt_stack/salt/lambda-dev/files/makedomainenv (hunk 5 above)
Comment (movestill) [reply salt_stack/salt/lambda-dev/files/makedomainenv]: Yeah, that would be a lot less lines of code!

**Pull request merged**